### ISSUE TITLE
[bitnami/keycloak] new ports for metrics and health endpoints

### ISSUE THREAD

**Issue by oberlon** [bitnami/charts#28630]
### Name and Version

bitnami/keycloak 22.0.0

### What architecture are you using?

amd64

### What steps will reproduce the bug?

Keycloak changed the ports for metrics and health ([https://www.keycloak.org/docs/25.0.0/release_notes/#management-port-for-metrics-and-health-endpoints](https://www.keycloak.org/docs/25.0.0/release_notes/#management-port-for-metrics-and-health-endpoints))

### What is the expected behavior?

Get keycloak metrics

### What do you see instead?

404 page

**Comment by ikatlinsky** [comment]
It would be nice to have a native solution in a helm chart to access metrics and health endpoints, currently, there are 2 options (as far as I can see):
- use `KC_LEGACY_OBSERVABILITY_INTERFACE` variable to rollback to keycloak <= 24 behavior 
- manually configure additional ports and mappings for pod/service/ingress

**Comment by javsalgar** [comment]
Hi!

Thank you so much for the feature request. I will forward it to the team but as it is not a critical feature we cannot guarantee an ETA. If you want to speed up the process, you can contribute with a PR and the team will check it.

**Comment by krankkkk** [comment]
For the next guy on how to enable metrics the mean time with the networkPolicy enabled:


```
  extraContainerPorts:
    - name: metrics
      containerPort: 9000
      protocol: TCP

  extraEnvVars:
    - name: KC_METRICS_ENABLED
      value: "true"
    #- name: KC_HTTP_METRICS_HISTOGRAMS_ENABLED
    #  value: "true"
    #- name: KC_CACHE_METRICS_HISTOGRAMS_ENABLED
    #  value: "true"

  metrics:
    enabled: true
    service:
      extraPorts:
        - name: metrics
          port: 9000
          protocol: TCP
          targetPort: metrics
    serviceMonitor:
      enabled: true
      port: metrics
      endpoints:
        - path: '{{ include "keycloak.httpPath" . }}metrics'
    prometheusRule:
      enabled: true

  networkPolicy:
    extraIngress:
    - from:
      - namespaceSelector:
          matchLabels:
            kubernetes.io/metadata.name: <your prometheus-namespace>
        podSelector:
          matchLabels:
            app.kubernetes.io/name: prometheus
      ports:
        - protocol: TCP
          port: 9000

```

**Comment by nix-power** [comment]
Did not work for me. I can access kc internal service on port 9000, but while trying to see `/metrics` endpoint get an empty reply from the server.
`curl --head -fsS http://idp-keycloak-metrics:9000/metrics
curl: (52) Empty reply from server`

**Comment by krankkkk** [comment]
@nix-power then you are probably running a older keycloak version, starting with 24 it is not served on the standart web server anymore, but instead on the management port (hence this ticket)

**Comment by nix-power** [comment]
Finally, everything is functioning with the following configuration, though in a somewhat unexpected way. The issue I previously encountered was due to the fact that my main Keycloak server runs as `HTTPS` by default. According to the [documentation](https://www.keycloak.org/server/management-interface), the management `HTTP`  server also operates over `HTTPS`, as clearly indicated in the log line I shared in my earlier comment. 

The configuration below made the  job:

```
  extraContainerPorts:
    - name: metrics
      containerPort: 9000
      protocol: TCP

  extraEnvVars:
    - name: KC_HTTP_MANAGEMENT_RELATIVE_PATH
      value: '/management'
    - name: KEYCLOAK_ENABLE_HEALTH_ENDPOINTS
      value: "true"
    - name: KEYCLOAK_ENABLE_STATISTICS
      value: "true"
    - name: KC_HOSTNAME_BACKCHANNEL_DYNAMIC
      value: "true"

  metrics:
    enabled: true
    service:
      annotations: null
      extraPorts:
        - name: metrics
          port: 9000
          protocol: TCP
          targetPort: metrics
    serviceMonitor:
      enabled: true
      port: metrics
      endpoints:
        - path: '{{ include "keycloak.httpPath" . }}/management/metrics'
        - path: '{{ include "keycloak.httpPath" . }}/management/realms/master/metrics'

   networkPolicy:
    enabled: true
    allowExternal: false
    allowExternalEgress: true
        extraIngress:
      - ports:
          - port: 9000
            protocol: TCP
        from:
          - namespaceSelector:
              matchLabels:
                kubernetes.io/metadata.name: idp
            podSelector:
              matchLabels:
                app.kubernetes.io/role: internal-apps
          - namespaceSelector:
              matchLabels:
                kubernetes.io/metadata.name: victoria-metrics
            podSelector:
              matchLabels:
                app.kubernetes.io/name: vmagent
```

right now when i probe `metrics` or `health` endpoints through idp-keykloak-metrics service interface  using 'https` i can clearly observe  them

```
curl -sk https://idp-keycloak-metrics:9000/management/metrics

# TYPE agroal_destroy_count counter
# HELP agroal_destroy_count Number of destroyed connections.
agroal_destroy_count_total{datasource="default"} 2.0
# TYPE jvm_gc_concurrent_phase_time_seconds summary
# HELP jvm_gc_concurrent_phase_time_seconds Time spent in concurrent phase
jvm_gc_concurrent_phase_time_seconds_count{action="end of concurrent GC pause",cause="No GC",gc="G1 Concurrent GC"} 4.0
jvm_gc_concurrent_phase_time_seconds_sum{action="end of concurrent GC pause",cause="No GC",gc="G1 Concurrent GC"} 0.036
# TYPE jvm_gc_concurrent_phase_time_seconds_max gauge
# HELP jvm_gc_concurrent_phase_time_seconds_max Time spent in concurrent phase
jvm_gc_concurrent_phase_time_seconds_max{action="end of concurrent GC pause",cause="No GC",gc="G1 Concurrent GC"} 0.0
# TYPE worker_pool_idle gauge
# HELP worker_pool_idle The number of resources from the pool currently used
worker_pool_idle{pool_name="vert.x-internal-blocking",pool_type="worker"} 20.0
worker_pool_idle{pool_name="vert.x-worker-thread",pool_type="worker"} 49.0
# TYPE vendor_jgroups_tcp_get_num_bytes_received gauge
# HELP vendor_jgroups_tcp_get_num_bytes_received Total number of bytes received (unicast + multicast)
vendor_jgroups_tcp_get_num_bytes_received{cache_manager="keycloak",cluster="ISPN",node="idp-keycloak-1-36472"} 952339.0
# TYPE vendor_cache_container_stats_number_of_entries gauge
# HELP vendor_cache_container_stats_number_of_entries Cache container total number of entries currently in all caches from this cache container
vendor_cache_container_stats_number_of_entries{cache_manager="keycloak",node="idp-keycloak-1-36472"} 339.0
```

```
curl -sk https://idp-keycloak-metrics:9000/management/health/ready
{
    "status": "UP",
    "checks": [
        {
            "name": "Keycloak database connections async health check",
            "status": "UP"
        }
    ]
}
```

The only unexpected thing there that endpoint of `/management/realms/master/metrics`  is not working as expected, through management interface, but it is still exposed via main `HTTP Server` of keycloak and main keycloak service

Not exposed (expdeted to expose)
```
curl -w '\n' -sk -X GET  https://idp-keycloak-metrics:9000/management/realms/master/metrics

<html><body><h1>Resource not found</h1></body></html>
```

Exposed (not expected to expose)
```
 curl -w '\n' -sk -X GET  http://idp-keycloak:8080/realms/master/metrics
# HELP keycloak_user_event_IMPERSONATE_total Generic KeyCloak User event
# TYPE keycloak_user_event_IMPERSONATE_total counter
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 87.24
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.724275111873E9
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 318.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1048576.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 5.94395136E9
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 6.8728832E8
```



But it looks more like a but in keycloak software itself and in management interface implementation that exposes only `/metrics` endpoint

**Comment by gabriel-suela** [comment]
i cant access the /realms/master/metrics endpoint on the default config but /metrics is fine. It is really suppose to have 2 metrics endpoints? /metrics dont have all of them?

**Comment by jazzlyn** [comment]
this is not solved. without additional config this triggers alerting.

**Comment by ikatlinsky** [comment]
Chart native solution is needed

**Comment by maschmi** [comment]
I'm also favouring a native solution. This costs time to fix and not having much time to deal with such issues is a reason I use this chart.

**Comment by javsalgar** [comment]
Thank you so much for the PR!

**Comment by maschmi** [comment]
@javsalgar It is a rather small and simple change. You can check it and weigh into the discussion if thats enough or not if you like.

**Comment by ikatlinsky** [comment]
> @javsalgar It is a rather small and simple change. You can check it and weigh into the discussion if thats enough or not if you like.

I am pretty sure that PR does not solve the problem:
- keycloak container should expose port 9000, which requires an update of the values file and statefulset.yaml file (https://github.com/bitnami/charts/blob/main/bitnami/keycloak/templates/statefulset.yaml#L243)
- metrics settings should also be updated, as currently, they are using the default keycloak port (https://github.com/bitnami/charts/blob/main/bitnami/keycloak/values.yaml#L1002)
- metrics service should also be updated to use correct pod port (https://github.com/bitnami/charts/blob/main/bitnami/keycloak/templates/metrics-service.yaml)

**Comment by maschmi** [comment]
Thank you for pointing this out. I will try and get at it over the next couple of days.

**Comment by maschmi** [comment]
@ikatlinsky Thank you again for pointing out the issues with my way to naive idea.  I've done the changes you suggested and also tested them as good as I was able.

I've decided to introduce a new value 'containerPorts.metrics' which default to 9000 but can be set to 8080 when soemone  runs with KC_LEGACY_OBSERVABILITY_INTERFACE. Only if the metrics port differs from the http port and metrics are enable the stateful set and the network policies are changed.

Further, I did not change the keycloak-metrics port from 8080 and only changed the targt of it. This should help making this change uninstrusive.
